python/telemetry/snet.py

```diff
@@ -168,10 +100,91 @@ def _decode(self, obj, context, path=None):
     'TargetData_ControlIsActive' / Int8ul  # flag, really
     )
 
+LTUFrameHeader = BitStruct(
+    'SrcId' / BitsInteger(7),
+    'DstId' / BitsInteger(7),
+    'FrCntTx' / BitsInteger(4),
+    'FrCntRx' / BitsInteger(4),
+    'SNR' / BitsInteger(4),
+    'AiTypeSrc' / BitsInteger(4),
+    'AiTypeDst' / BitsInteger(4),
+    'DfcId' / BitsInteger(2),
+    'Caller' / Flag,
+    'Arq' / Flag,
+    'PduTypeId' / Flag,
+    'BchRq' / Flag,
+    'Hailing' / Flag,
+    'UdFl1' / Flag,
+    'PduLength' / BitsInteger(10),
+    'CRC13' / BitsInteger(13),
+    'CRC5' / BitsInteger(5),
+    Padding(2)
+    )
+
+# For reference:
+#
+# To send mailbox messages
+# FCIDMajor = 56  (COM0)
+# FCIDSub = 34 || 134  (COM1)
+
+MailboxTelemetry = Struct(
+    'SysTime' / BitsInteger(4),
+    'MemDataType' / BitsInteger(1),
+    'MemData32' / Hex(GreedyBytes)
+    # 'MemData32' / LinearAdapter(32, Int8ub)
+    )
+
 
-snet = Struct(
-    'header' / SNETFrameHeader,
-    'telemetry' / Switch(lambda c: (c.header.FCIDMajor, c.header.FCIDSub), {
+class TimeAdapter(Adapter):
+    def _encode(self, obj, context, path=None):
+        d = int((obj - datetime.datetime(2000, 1, 1))*2)
+        return Container(
+            days=d.days,
+            milliseconds=d.seconds * 1000 + d.microseconds / 1000)
+
+    def _decode(self, obj, context, path=None):
+        return (datetime.datetime(2000, 1, 1)
+                + datetime.timedelta(seconds=float(obj)/2.0))
+
+
+TimeStamp = TimeAdapter(BitsInteger(32, swapped=True))
+
+PDUFrameHeaderExtension = BitStruct(
+    'VersNo' / BitsInteger(2),
+    'DFCID' / BitsInteger(2),
+    'ExtensionRFU' / BitsInteger(4),
+    'ChannelInfo' / BitsInteger(8),
+    'QoS' / Flag,
+    'PDUTypeID' / Flag,
+    'ARQ' / Flag,
+    'ControlRFU' / BitsInteger(5),
+    'TimeTagSub' / BitsInteger(16),
+    'SCID' / BitsInteger(10),
+    'SeqNo' / BitsInteger(14)
+    )
+
+PDUFrameHeader = BitStruct(
+    Const(0b111100110101000000, BitsInteger(18)),
+    'CRC' / BitsInteger(14),
+    'FCIDMajor' / BitsInteger(6),
+    'FCIDSub' / BitsInteger(10),
+    'Urgent' / Flag,
+    'Extended' / Flag,
+    'CheckCRC' / Flag,
+    'Multiframe' / Flag,
+    'TimeTaggedSetting' / Flag,
+    'TimeTagged' / Flag,
+    'DataLength' / BitsInteger(10),
+    'TimeTag' / If(lambda c: c.TimeTagged, TimeStamp),
+    'Extension' / If(lambda c: c.Extended, PDUFrameHeaderExtension),
+    'UserData' / Switch(lambda c: (c.FCIDMajor, c.FCIDSub), {
         (0, 0): ADCSTelemetry,
         (9, 0): EPSTelemetry,
-    }, default=Pass))
+        (54, 17): MailboxTelemetry,  # COM0
+        (54, 117): MailboxTelemetry  # COM1
+    },
+      # why parse instead of PASS? it would be good to have a repesentation
+      # in case the wave file contains a new pair of FCIDMajor and FCIDSub not
+      # supported. It creates a bit of lag, but I guess is worth it instead of
+      # losing the UserData.
```

Comment:
The goal of a telemetry parser is to produce human readable output. All the possible telemetry formats that the satellite can transmit are supposed to be implemented in this file. For other satellites, often this is not possible due to lack of documentation or development effort, in which case sometimes a greedy bytes or similar is used for the payload of the frame.

Data should not be lost if a frame cannot be parsed, since the original data can be written to a KISS file.

Reply:
I agree. Is this comment different from what I put in the file?

I would like to extend my comment pointing the details you mention, but I want to make sure if I am reading your comment correctly.

Reply:
I think your comment matches your code, but my point is that:
1. The telemetry parser should probably support all the different types of frames transmitted by the satellite, so there wouldn't be unknown frames.
2. Documentation for all the frame types should generally be publicly available as per item 6.1 in [IARU Satellite Frequency Coordination request](https://www.iaru.org/wp-content/uploads/2020/06/iaru_amateur_satellite_coordination_request_v40.doc).
3. In the event that a payload doesn't match the formats supported by the telemetry parser, the parser should just ignore the payload (as it does now), since its goal is to parse fields into human readable format, and it cannot do so for frames of unsupported format.

Reply:
oh, I didn't know that, and now it makes sense why it was set to `default=Pass` in the first place.

I agree, and thanks for the information/context.